// @xct007/frieren-scraper/dist/cjs/Constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIsItsRose = exports.BokepSinBaseUrl = exports.PrivateWorkerApiItsrose = exports.PrivateApiPhotoFunia = exports.PinterestDownloadBaseUrls = exports.ApkmodyIoBaseUrl = exports.MusicApiJamendoBaseUrl = exports.FacebookDownloadBaseUrls = exports.DanbooruBaseUrl = exports.TiktokDownloadRapidApiServer = exports.InstagramDownloadBaseUrlV1 = exports.UnsplashBaseUrl = exports.StatusWaIndonesiaBaseUrl = exports.AnoboyBaseUrl = exports.KomikuIdBaseUrl = exports.YoutubeDownloadBaseUrl = exports.YoutubeSearchBaseUrl = exports.OtakudesuBaseUrl = exports.DoujindesuBaseUrl = void 0;
exports.DoujindesuBaseUrl = "https://212.32.226.234";
exports.OtakudesuBaseUrl = "https://otakudesu.ltd";
exports.YoutubeSearchBaseUrl = "https://youtube.com";
exports.YoutubeDownloadBaseUrl = "https://onlinevideoconverter.pro";
exports.KomikuIdBaseUrl = "https://komiku.id";
exports.AnoboyBaseUrl = "https://anoboy.ink";
exports.StatusWaIndonesiaBaseUrl = "https://www.videostatusmarket.com";
exports.UnsplashBaseUrl = "https://api.unsplash.com";
exports.InstagramDownloadBaseUrlV1 = "https://downloadgram.org";
exports.TiktokDownloadRapidApiServer = "https://tiktok-video-no-watermark2.p.rapidapi.com";
exports.DanbooruBaseUrl = "https://danbooru.donmai.us";
exports.FacebookDownloadBaseUrls = {
    v1: "https://getmyfb.com",
};
exports.MusicApiJamendoBaseUrl = "https://api.jamendo.com";
exports.ApkmodyIoBaseUrl = "https://apkmody.io";
exports.PinterestDownloadBaseUrls = {
    v1: "https://pinterestvideo.com",
};
exports.PrivateApiPhotoFunia = "https://photofunia.itsrose.site";
exports.PrivateWorkerApiItsrose = "https://worker.itsrose.site";
exports.BokepSinBaseUrl = "https://bokepsin.guru";
exports.APIsItsRose = "https://api.itsrose.site";
