// youtubei.js/dist/src/parser/classes/comments/CommentDialog.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = __importDefault(require("../.."));
const Text_1 = __importDefault(require("../misc/Text"));
const Thumbnail_1 = __importDefault(require("../misc/Thumbnail"));
const helpers_1 = require("../../helpers");
class CommentDialog extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.editable_text = new Text_1.default(data.editableText);
        this.author_thumbnail = Thumbnail_1.default.fromResponse(data.authorThumbnail);
        this.submit_button = __1.default.parseItem(data.submitButton);
        this.cancel_button = __1.default.parseItem(data.cancelButton);
        this.placeholder = new Text_1.default(data.placeholderText);
        this.emoji_button = __1.default.parseItem(data.emojiButton);
        this.emoji_picker = __1.default.parseItem(data.emojiPicker);
    }
}
CommentDialog.type = 'CommentDialog';
exports.default = CommentDialog;
