// youtubei.js/dist/src/parser/classes/comments/CreatorHeart.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../../helpers");
const Thumbnail_1 = __importDefault(require("../misc/Thumbnail"));
class CreatorHeart extends helpers_1.YTNode {
    constructor(data) {
        var _a, _b, _c;
        super();
        this.creator_thumbnail = Thumbnail_1.default.fromResponse(data.creatorThumbnail);
        this.heart_icon_type = (_a = data.heartIcon) === null || _a === void 0 ? void 0 : _a.iconType;
        this.heart_color = {
            basic_color_palette_data: {
                foreground_title_color: (_c = (_b = data.heartColor) === null || _b === void 0 ? void 0 : _b.basicColorPaletteData) === null || _c === void 0 ? void 0 : _c.foregroundTitleColor
            }
        };
        this.hearted_tooltip = data.heartedTooltip;
        this.is_hearted = data.isHearted;
        this.is_enabled = data.isEnabled;
        this.kennedy_heart_color_string = data.kennedyHeartColorString;
    }
}
CreatorHeart.type = 'CreatorHeart';
exports.default = CreatorHeart;
