// @xct007/frieren-scraper/dist/cjs/downloader/statusWaIndonesia.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.search = exports.popular = void 0;
const Utils_1 = require("../Utils");
const Constant_1 = require("../Constant");
async function popular(page = "1", seed = "6316") {
    try {
        const { data } = await Utils_1.Axios.request({
            url: Constant_1.StatusWaIndonesiaBaseUrl +
                "/videostatus_studio/videostatus_indonesia/get_new_video_portrait.php",
            method: "POST",
            headers: { ["Content-Type"]: "application/x-www-form-urlencoded" },
            data: new URLSearchParams({ seed, page, type: "popular" }),
        }).catch((e) => e?.response);
        if (data && typeof data === "object") {
            return data.items;
        }
        else {
            throw new Error(`data: ${typeof data}`);
        }
    }
    catch (e) {
        return {
            error: true,
            message: String(e),
        };
    }
}
exports.popular = popular;
async function search(query, page = "1", seed = "3013") {
    try {
        const { data } = await Utils_1.Axios.request({
            url: Constant_1.StatusWaIndonesiaBaseUrl +
                "/videostatus_studio/videostatus_indonesia/get_new_video_portrait.php",
            method: "POST",
            headers: { ["Content-Type"]: "application/x-www-form-urlencoded" },
            data: new URLSearchParams({ s: query, seed, page, type: "search" }),
        }).catch((e) => e?.response);
        if (data && typeof data === "object") {
            return data.items;
        }
        else {
            throw new Error(`data: ${typeof data}`);
        }
    }
    catch (e) {
        return {
            error: true,
            message: String(e),
        };
    }
}
exports.search = search;
