// @xct007/frieren-scraper/dist/esm/Constant.js
export const DoujindesuBaseUrl = "https://212.32.226.234";
export const OtakudesuBaseUrl = "https://otakudesu.ltd";
export const YoutubeSearchBaseUrl = "https://youtube.com";
export const YoutubeDownloadBaseUrl = "https://onlinevideoconverter.pro";
export const KomikuIdBaseUrl = "https://komiku.id";
export const AnoboyBaseUrl = "https://anoboy.ink";
export const StatusWaIndonesiaBaseUrl = "https://www.videostatusmarket.com";
export const UnsplashBaseUrl = "https://api.unsplash.com";
export const InstagramDownloadBaseUrlV1 = "https://downloadgram.org";
export const TiktokDownloadRapidApiServer = "https://tiktok-video-no-watermark2.p.rapidapi.com";
export const DanbooruBaseUrl = "https://danbooru.donmai.us";
export const FacebookDownloadBaseUrls = {
    v1: "https://getmyfb.com",
};
export const MusicApiJamendoBaseUrl = "https://api.jamendo.com";
export const ApkmodyIoBaseUrl = "https://apkmody.io";
export const PinterestDownloadBaseUrls = {
    v1: "https://pinterestvideo.com",
};
export const PrivateApiPhotoFunia = "https://photofunia.itsrose.site";
export const PrivateWorkerApiItsrose = "https://worker.itsrose.site";
export const BokepSinBaseUrl = "https://bokepsin.guru";
export const APIsItsRose = "https://api.itsrose.site";
