// youtubei.js/dist/src/parser/ytmusic/Library.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Library_page, _Library_actions, _Library_continuation, _LibraryContinuation_page, _LibraryContinuation_actions, _LibraryContinuation_continuation;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LibraryContinuation = void 0;
const __1 = __importStar(require(".."));
const Grid_1 = __importDefault(require("../classes/Grid"));
const MusicShelf_1 = __importDefault(require("../classes/MusicShelf"));
const MusicSideAlignedItem_1 = __importDefault(require("../classes/MusicSideAlignedItem"));
const NavigationEndpoint_1 = __importDefault(require("../classes/NavigationEndpoint"));
const SectionList_1 = __importDefault(require("../classes/SectionList"));
const ChipCloud_1 = __importDefault(require("../classes/ChipCloud"));
const ChipCloudChip_1 = __importDefault(require("../classes/ChipCloudChip"));
const MusicMultiSelectMenuItem_1 = __importDefault(require("../classes/menus/MusicMultiSelectMenuItem"));
const MusicSortFilterButton_1 = __importDefault(require("../classes/MusicSortFilterButton"));
const Utils_1 = require("../../utils/Utils");
class Library {
    constructor(response, actions) {
        var _a, _b, _c, _d;
        _Library_page.set(this, void 0);
        _Library_actions.set(this, void 0);
        _Library_continuation.set(this, void 0);
        __classPrivateFieldSet(this, _Library_page, __1.default.parseResponse(response.data), "f");
        __classPrivateFieldSet(this, _Library_actions, actions, "f");
        const section_list = __classPrivateFieldGet(this, _Library_page, "f").contents_memo.getType(SectionList_1.default).first();
        this.header = (_a = section_list === null || section_list === void 0 ? void 0 : section_list.header) === null || _a === void 0 ? void 0 : _a.item().as(MusicSideAlignedItem_1.default);
        this.contents = (_b = section_list === null || section_list === void 0 ? void 0 : section_list.contents) === null || _b === void 0 ? void 0 : _b.as(Grid_1.default, MusicShelf_1.default);
        __classPrivateFieldSet(this, _Library_continuation, (_d = (_c = this.contents) === null || _c === void 0 ? void 0 : _c.find((list) => list.continuation)) === null || _d === void 0 ? void 0 : _d.continuation, "f");
    }
    /**
     * Applies given sort option to the library items.
     */
    applySort(sort_by) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        return __awaiter(this, void 0, void 0, function* () {
            let target_item;
            if (typeof sort_by === 'string') {
                const button = (_a = __classPrivateFieldGet(this, _Library_page, "f").contents_memo.getType(MusicSortFilterButton_1.default)) === null || _a === void 0 ? void 0 : _a[0];
                const options = (_b = button.menu) === null || _b === void 0 ? void 0 : _b.options.filter((item) => item instanceof MusicMultiSelectMenuItem_1.default);
                target_item = options === null || options === void 0 ? void 0 : options.find((item) => item.title === sort_by);
                if (!target_item)
                    throw new Utils_1.InnertubeError(`Sort option "${sort_by}" not found`, { available_filters: options.map((item) => item.title) });
            }
            else if (sort_by instanceof MusicMultiSelectMenuItem_1.default) {
                target_item = sort_by;
            }
            if (!target_item)
                throw new Utils_1.InnertubeError('Invalid sort option');
            if (target_item.selected)
                return this;
            const cmd = (_f = (_e = (_d = (_c = target_item.endpoint) === null || _c === void 0 ? void 0 : _c.payload) === null || _d === void 0 ? void 0 : _d.commands) === null || _e === void 0 ? void 0 : _e.find((cmd) => cmd.browseSectionListReloadEndpoint)) === null || _f === void 0 ? void 0 : _f.browseSectionListReloadEndpoint;
            if (!cmd)
                throw new Utils_1.InnertubeError('Failed to find sort option command');
            const response = yield __classPrivateFieldGet(this, _Library_actions, "f").execute('/browse', {
                client: 'YTMUSIC',
                continuation: cmd.continuation.reloadContinuationData.continuation,
                parse: true
            });
            const previously_selected_item = (_g = __classPrivateFieldGet(this, _Library_page, "f").contents_memo.getType(MusicMultiSelectMenuItem_1.default)) === null || _g === void 0 ? void 0 : _g.find((item) => item.selected);
            if (previously_selected_item)
                previously_selected_item.selected = false;
            target_item.selected = true;
            this.contents = (_j = (_h = response.continuation_contents) === null || _h === void 0 ? void 0 : _h.as(__1.SectionListContinuation).contents) === null || _j === void 0 ? void 0 : _j.as(Grid_1.default, MusicShelf_1.default);
            return this;
        });
    }
    /**
     * Applies given filter to the library.
     */
    applyFilter(filter) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            let target_chip;
            const chip_cloud = (_a = __classPrivateFieldGet(this, _Library_page, "f").contents_memo.getType(ChipCloud_1.default)) === null || _a === void 0 ? void 0 : _a[0];
            if (typeof filter === 'string') {
                target_chip = chip_cloud.chips.get({ text: filter });
                if (!target_chip)
                    throw new Utils_1.InnertubeError(`Filter "${filter}" not found`, { available_filters: this.filters });
            }
            else if (filter instanceof ChipCloudChip_1.default) {
                target_chip = filter;
            }
            if (!target_chip)
                throw new Utils_1.InnertubeError('Invalid filter', filter);
            const target_cmd = new NavigationEndpoint_1.default((_d = (_c = (_b = target_chip.endpoint) === null || _b === void 0 ? void 0 : _b.payload) === null || _c === void 0 ? void 0 : _c.commands) === null || _d === void 0 ? void 0 : _d[0]);
            const response = yield target_cmd.call(__classPrivateFieldGet(this, _Library_actions, "f"), { client: 'YTMUSIC' });
            return new Library(response, __classPrivateFieldGet(this, _Library_actions, "f"));
        });
    }
    /**
     * Retrieves continuation of the library items.
     */
    getContinuation() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _Library_continuation, "f"))
                throw new Utils_1.InnertubeError('No continuation available');
            const page = yield __classPrivateFieldGet(this, _Library_actions, "f").execute('/browse', {
                client: 'YTMUSIC',
                continuation: __classPrivateFieldGet(this, _Library_continuation, "f")
            });
            return new LibraryContinuation(page, __classPrivateFieldGet(this, _Library_actions, "f"));
        });
    }
    get has_continuation() {
        return !!__classPrivateFieldGet(this, _Library_continuation, "f");
    }
    get sort_options() {
        var _a, _b;
        const button = (_a = __classPrivateFieldGet(this, _Library_page, "f").contents_memo.getType(MusicSortFilterButton_1.default)) === null || _a === void 0 ? void 0 : _a[0];
        const options = (_b = button.menu) === null || _b === void 0 ? void 0 : _b.options.filter((item) => item instanceof MusicMultiSelectMenuItem_1.default);
        return options.map((item) => item.title);
    }
    get filters() {
        var _a, _b;
        return ((_b = (_a = __classPrivateFieldGet(this, _Library_page, "f").contents_memo.getType(ChipCloud_1.default)) === null || _a === void 0 ? void 0 : _a.first()) === null || _b === void 0 ? void 0 : _b.chips.map((chip) => chip.text)) || [];
    }
    get page() {
        return __classPrivateFieldGet(this, _Library_page, "f");
    }
}
_Library_page = new WeakMap(), _Library_actions = new WeakMap(), _Library_continuation = new WeakMap();
class LibraryContinuation {
    constructor(response, actions) {
        var _a, _b, _c, _d, _e;
        _LibraryContinuation_page.set(this, void 0);
        _LibraryContinuation_actions.set(this, void 0);
        _LibraryContinuation_continuation.set(this, void 0);
        __classPrivateFieldSet(this, _LibraryContinuation_page, __1.default.parseResponse(response.data), "f");
        __classPrivateFieldSet(this, _LibraryContinuation_actions, actions, "f");
        this.contents = ((_a = __classPrivateFieldGet(this, _LibraryContinuation_page, "f").continuation_contents) === null || _a === void 0 ? void 0 : _a.hasKey('contents'))
            ? (_b = __classPrivateFieldGet(this, _LibraryContinuation_page, "f").continuation_contents) === null || _b === void 0 ? void 0 : _b.key('contents').array() :
            (_c = __classPrivateFieldGet(this, _LibraryContinuation_page, "f").continuation_contents) === null || _c === void 0 ? void 0 : _c.key('items').array();
        __classPrivateFieldSet(this, _LibraryContinuation_continuation, ((_d = __classPrivateFieldGet(this, _LibraryContinuation_page, "f").continuation_contents) === null || _d === void 0 ? void 0 : _d.key('continuation').isNull())
            ? null : (_e = __classPrivateFieldGet(this, _LibraryContinuation_page, "f").continuation_contents) === null || _e === void 0 ? void 0 : _e.key('continuation').string(), "f");
    }
    getContinuation() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _LibraryContinuation_continuation, "f"))
                throw new Utils_1.InnertubeError('No continuation available');
            const page = yield __classPrivateFieldGet(this, _LibraryContinuation_actions, "f").execute('/browse', {
                client: 'YTMUSIC',
                continuation: __classPrivateFieldGet(this, _LibraryContinuation_continuation, "f")
            });
            return new LibraryContinuation(page, __classPrivateFieldGet(this, _LibraryContinuation_actions, "f"));
        });
    }
    get has_continuation() {
        return !!__classPrivateFieldGet(this, _LibraryContinuation_continuation, "f");
    }
    get page() {
        return __classPrivateFieldGet(this, _LibraryContinuation_page, "f");
    }
}
exports.LibraryContinuation = LibraryContinuation;
_LibraryContinuation_page = new WeakMap(), _LibraryContinuation_actions = new WeakMap(), _LibraryContinuation_continuation = new WeakMap();
exports.default = Library;
