// youtubei.js/dist/src/parser/classes/C4TabbedHeader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = __importDefault(require("../index"));
const Author_1 = __importDefault(require("./misc/Author"));
const Text_1 = __importDefault(require("./misc/Text"));
const Thumbnail_1 = __importDefault(require("./misc/Thumbnail"));
const helpers_1 = require("../helpers");
class C4TabbedHeader extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.author = new Author_1.default({
            simpleText: data.title,
            navigationEndpoint: data.navigationEndpoint
        }, data.badges, data.avatar);
        this.banner = Thumbnail_1.default.fromResponse(data.banner);
        this.tv_banner = Thumbnail_1.default.fromResponse(data.tvBanner);
        this.mobile_banner = Thumbnail_1.default.fromResponse(data.mobileBanner);
        this.subscribers = new Text_1.default(data.subscriberCountText);
        this.videos_count = new Text_1.default(data.videosCountText);
        this.sponsor_button = index_1.default.parseItem(data.sponsorButton);
        this.subscribe_button = index_1.default.parseItem(data.subscribeButton);
        this.header_links = index_1.default.parseItem(data.headerLinks);
        this.channel_handle = new Text_1.default(data.channelHandleText);
        this.channel_id = data.channelId;
    }
}
C4TabbedHeader.type = 'C4TabbedHeader';
exports.default = C4TabbedHeader;
