// youtubei.js/dist/src/parser/classes/ExpandableMetadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = __importDefault(require(".."));
const Text_1 = __importDefault(require("./misc/Text"));
const Thumbnail_1 = __importDefault(require("./misc/Thumbnail"));
const helpers_1 = require("../helpers");
class ExpandableMetadata extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.header = {
            collapsed_title: new Text_1.default(data.header.collapsedTitle),
            collapsed_thumbnail: Thumbnail_1.default.fromResponse(data.header.collapsedThumbnail),
            collapsed_label: new Text_1.default(data.header.collapsedLabel),
            expanded_title: new Text_1.default(data.header.expandedTitle)
        };
        this.expanded_content = __1.default.parseItem(data.expandedContent);
        this.expand_button = __1.default.parseItem(data.expandButton);
        this.collapse_button = __1.default.parseItem(data.collapseButton);
    }
}
ExpandableMetadata.type = 'ExpandableMetadata';
exports.default = ExpandableMetadata;
