// youtubei.js/dist/src/parser/classes/InteractiveTabbedHeader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = __importDefault(require(".."));
const helpers_1 = require("../helpers");
const Text_1 = __importDefault(require("./misc/Text"));
const Thumbnail_1 = __importDefault(require("./misc/Thumbnail"));
const SubscribeButton_1 = __importDefault(require("./SubscribeButton"));
const MetadataBadge_1 = __importDefault(require("./MetadataBadge"));
const Button_1 = __importDefault(require("./Button"));
class InteractiveTabbedHeader extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.header_type = data.type;
        this.title = new Text_1.default(data.title);
        this.description = new Text_1.default(data.description);
        this.metadata = new Text_1.default(data.metadata);
        this.badges = __1.default.parseArray(data.badges, MetadataBadge_1.default);
        this.box_art = Thumbnail_1.default.fromResponse(data.boxArt);
        this.banner = Thumbnail_1.default.fromResponse(data.banner);
        this.buttons = __1.default.parseArray(data.buttons, [SubscribeButton_1.default, Button_1.default]);
        this.auto_generated = new Text_1.default(data.autoGenerated);
    }
}
InteractiveTabbedHeader.type = 'InteractiveTabbedHeader';
exports.default = InteractiveTabbedHeader;
