// youtubei.js/dist/src/parser/classes/MacroMarkersListItem.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Text_1 = __importDefault(require("./misc/Text"));
const Thumbnail_1 = __importDefault(require("./misc/Thumbnail"));
const NavigationEndpoint_1 = __importDefault(require("./NavigationEndpoint"));
const helpers_1 = require("../helpers");
class MacroMarkersListItem extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.title = new Text_1.default(data.title);
        this.time_description = new Text_1.default(data.timeDescription);
        this.thumbnail = Thumbnail_1.default.fromResponse(data.thumbnail);
        this.on_tap_endpoint = new NavigationEndpoint_1.default(data.onTap);
        this.layout = data.layout;
        this.is_highlighted = data.isHighlighted;
    }
}
MacroMarkersListItem.type = 'MacroMarkersListItem';
exports.default = MacroMarkersListItem;
