// youtubei.js/dist/src/parser/classes/livechat/items/LiveChatTickerPaidStickerItem.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = __importDefault(require("../../../index"));
const LiveChatAuthorBadge_1 = __importDefault(require("../../LiveChatAuthorBadge"));
const MetadataBadge_1 = __importDefault(require("../../MetadataBadge"));
const Text_1 = __importDefault(require("../../misc/Text"));
const Thumbnail_1 = __importDefault(require("../../misc/Thumbnail"));
const NavigationEndpoint_1 = __importDefault(require("../../NavigationEndpoint"));
const helpers_1 = require("../../../helpers");
class LiveChatTickerPaidStickerItem extends helpers_1.YTNode {
    constructor(data) {
        var _a, _b;
        super();
        this.author = {
            id: data.authorExternalChannelId,
            name: new Text_1.default(data === null || data === void 0 ? void 0 : data.authorName),
            thumbnails: Thumbnail_1.default.fromResponse(data.authorPhoto),
            badges: (0, helpers_1.observe)([]).as(LiveChatAuthorBadge_1.default, MetadataBadge_1.default),
            is_moderator: null,
            is_verified: null,
            is_verified_artist: null
        };
        const badges = index_1.default.parseArray(data.authorBadges, [MetadataBadge_1.default, LiveChatAuthorBadge_1.default]);
        this.author.badges = badges;
        this.author.is_moderator = badges ? badges.some((badge) => badge.icon_type == 'MODERATOR') : null;
        this.author.is_verified = badges ? badges.some((badge) => badge.style == 'BADGE_STYLE_TYPE_VERIFIED') : null;
        this.author.is_verified_artist = badges ? badges.some((badge) => badge.style == 'BADGE_STYLE_TYPE_VERIFIED_ARTIST') : null;
        this.amount = new Text_1.default(data.amount);
        this.duration_sec = data.durationSec;
        this.full_duration_sec = data.fullDurationSec;
        this.show_item = index_1.default.parseItem((_b = (_a = data.showItemEndpoint) === null || _a === void 0 ? void 0 : _a.showLiveChatItemEndpoint) === null || _b === void 0 ? void 0 : _b.renderer);
        this.show_item_endpoint = new NavigationEndpoint_1.default(data.showItemEndpoint);
        this.id = data.id;
    }
}
LiveChatTickerPaidStickerItem.type = 'LiveChatTickerPaidStickerItem';
exports.default = LiveChatTickerPaidStickerItem;
