// youtubei.js/dist/src/parser/classes/livechat/items/LiveChatPaidSticker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../../../helpers");
const index_1 = __importDefault(require("../../../index"));
const LiveChatAuthorBadge_1 = __importDefault(require("../../LiveChatAuthorBadge"));
const MetadataBadge_1 = __importDefault(require("../../MetadataBadge"));
const Text_1 = __importDefault(require("../../misc/Text"));
const Thumbnail_1 = __importDefault(require("../../misc/Thumbnail"));
const NavigationEndpoint_1 = __importDefault(require("../../NavigationEndpoint"));
class LiveChatPaidSticker extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.id = data.id;
        this.author = {
            id: data.authorExternalChannelId,
            name: new Text_1.default(data.authorName),
            thumbnails: Thumbnail_1.default.fromResponse(data.authorPhoto),
            badges: (0, helpers_1.observe)([]).as(LiveChatAuthorBadge_1.default, MetadataBadge_1.default),
            is_moderator: null,
            is_verified: null,
            is_verified_artist: null
        };
        const badges = index_1.default.parseArray(data.authorBadges, [MetadataBadge_1.default, LiveChatAuthorBadge_1.default]);
        this.author.badges = badges;
        this.author.is_moderator = badges ? badges.some((badge) => badge.icon_type == 'MODERATOR') : null;
        this.author.is_verified = badges ? badges.some((badge) => badge.style == 'BADGE_STYLE_TYPE_VERIFIED') : null;
        this.author.is_verified_artist = badges ? badges.some((badge) => badge.style == 'BADGE_STYLE_TYPE_VERIFIED_ARTIST') : null;
        this.money_chip_background_color = data.moneyChipBackgroundColor;
        this.money_chip_text_color = data.moneyChipTextColor;
        this.background_color = data.backgroundColor;
        this.author_name_text_color = data.authorNameTextColor;
        this.sticker = Thumbnail_1.default.fromResponse(data.sticker);
        this.purchase_amount = new Text_1.default(data.purchaseAmountText).toString();
        this.menu_endpoint = new NavigationEndpoint_1.default(data.contextMenuEndpoint);
        this.context_menu = this.menu_endpoint;
        this.timestamp = Math.floor(parseInt(data.timestampUsec) / 1000);
    }
}
LiveChatPaidSticker.type = 'LiveChatPaidSticker';
exports.default = LiveChatPaidSticker;
