// youtubei.js/dist/src/parser/classes/livechat/items/LiveChatTickerSponsorItem.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = __importDefault(require("../../.."));
const helpers_1 = require("../../../helpers");
const LiveChatAuthorBadge_1 = __importDefault(require("../../LiveChatAuthorBadge"));
const MetadataBadge_1 = __importDefault(require("../../MetadataBadge"));
const Text_1 = __importDefault(require("../../misc/Text"));
const Thumbnail_1 = __importDefault(require("../../misc/Thumbnail"));
class LiveChatTickerSponsorItem extends helpers_1.YTNode {
    constructor(data) {
        super();
        this.id = data.id;
        this.detail = new Text_1.default(data.detailText);
        this.author = {
            id: data.authorExternalChannelId,
            name: new Text_1.default(data === null || data === void 0 ? void 0 : data.authorName),
            thumbnails: Thumbnail_1.default.fromResponse(data.sponsorPhoto),
            badges: (0, helpers_1.observe)([]).as(LiveChatAuthorBadge_1.default, MetadataBadge_1.default),
            is_moderator: null,
            is_verified: null,
            is_verified_artist: null
        };
        const badges = __1.default.parseArray(data.authorBadges, [MetadataBadge_1.default, LiveChatAuthorBadge_1.default]);
        this.author.badges = badges;
        this.author.is_moderator = badges ? badges.some((badge) => badge.icon_type == 'MODERATOR') : null;
        this.author.is_verified = badges ? badges.some((badge) => badge.style == 'BADGE_STYLE_TYPE_VERIFIED') : null;
        this.author.is_verified_artist = badges ? badges.some((badge) => badge.style == 'BADGE_STYLE_TYPE_VERIFIED_ARTIST') : null;
        this.duration_sec = data.durationSec;
        // TODO: finish this
    }
}
LiveChatTickerSponsorItem.type = 'LiveChatTickerSponsorItem';
exports.default = LiveChatTickerSponsorItem;
