// @xct007/frieren-scraper/dist/esm/downloader/statusWaIndonesia.js
import { Axios } from "../Utils.js";
import { StatusWaIndonesiaBaseUrl } from "../Constant.js";
async function popular(page = "1", seed = "6316") {
    try {
        const { data } = await Axios.request({
            url: StatusWaIndonesiaBaseUrl +
                "/videostatus_studio/videostatus_indonesia/get_new_video_portrait.php",
            method: "POST",
            headers: { ["Content-Type"]: "application/x-www-form-urlencoded" },
            data: new URLSearchParams({ seed, page, type: "popular" }),
        }).catch((e) => e?.response);
        if (data && typeof data === "object") {
            return data.items;
        }
        else {
            throw new Error(`data: ${typeof data}`);
        }
    }
    catch (e) {
        return {
            error: true,
            message: String(e),
        };
    }
}
async function search(query, page = "1", seed = "3013") {
    try {
        const { data } = await Axios.request({
            url: StatusWaIndonesiaBaseUrl +
                "/videostatus_studio/videostatus_indonesia/get_new_video_portrait.php",
            method: "POST",
            headers: { ["Content-Type"]: "application/x-www-form-urlencoded" },
            data: new URLSearchParams({ s: query, seed, page, type: "search" }),
        }).catch((e) => e?.response);
        if (data && typeof data === "object") {
            return data.items;
        }
        else {
            throw new Error(`data: ${typeof data}`);
        }
    }
    catch (e) {
        return {
            error: true,
            message: String(e),
        };
    }
}
export { popular, search };
