// youtubei.js/dist/src/parser/youtube/Search.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Feed_1 = __importDefault(require("../../core/Feed"));
const HorizontalCardList_1 = __importDefault(require("../classes/HorizontalCardList"));
const ItemSection_1 = __importDefault(require("../classes/ItemSection"));
const SearchRefinementCard_1 = __importDefault(require("../classes/SearchRefinementCard"));
const SectionList_1 = __importDefault(require("../classes/SectionList"));
const UniversalWatchCard_1 = __importDefault(require("../classes/UniversalWatchCard"));
const Utils_1 = require("../../utils/Utils");
class Search extends Feed_1.default {
    constructor(actions, data, already_parsed = false) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        super(actions, data, already_parsed);
        const contents = ((_b = (_a = this.page.contents_memo.getType(SectionList_1.default)) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.contents) ||
            ((_c = this.page.on_response_received_commands) === null || _c === void 0 ? void 0 : _c[0].contents);
        this.results = (_d = contents.firstOfType(ItemSection_1.default)) === null || _d === void 0 ? void 0 : _d.contents;
        this.refinements = this.page.refinements || [];
        this.estimated_results = this.page.estimated_results;
        this.watch_card = (_f = (_e = this.page) === null || _e === void 0 ? void 0 : _e.contents_memo.getType(UniversalWatchCard_1.default)) === null || _f === void 0 ? void 0 : _f[0];
        this.refinement_cards = (_h = (_g = this.results) === null || _g === void 0 ? void 0 : _g.get({ type: 'HorizontalCardList' }, true)) === null || _h === void 0 ? void 0 : _h.as(HorizontalCardList_1.default);
    }
    /**
     * Applies given refinement card and returns a new {@link Search} object. Use {@link refinement_card_queries} to get a list of available refinement cards.
     */
    selectRefinementCard(card) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            let target_card;
            if (typeof card === 'string') {
                if (!this.refinement_cards)
                    throw new Utils_1.InnertubeError('No refinement cards found.');
                target_card = (_b = (_a = this.refinement_cards) === null || _a === void 0 ? void 0 : _a.cards.get({ query: card })) === null || _b === void 0 ? void 0 : _b.as(SearchRefinementCard_1.default);
                if (!target_card)
                    throw new Utils_1.InnertubeError(`Refinement card "${card}" not found`, { available_cards: this.refinement_card_queries });
            }
            else if (card.type === 'SearchRefinementCard') {
                target_card = card;
            }
            else {
                throw new Utils_1.InnertubeError('Invalid refinement card!');
            }
            const page = yield target_card.endpoint.call(this.actions, { parse: true });
            return new Search(this.actions, page, true);
        });
    }
    /**
     * Returns a list of refinement card queries.
     */
    get refinement_card_queries() {
        var _a;
        return ((_a = this.refinement_cards) === null || _a === void 0 ? void 0 : _a.cards.as(SearchRefinementCard_1.default).map((card) => card.query)) || [];
    }
    /**
     * Retrieves next batch of results.
     */
    getContinuation() {
        return __awaiter(this, void 0, void 0, function* () {
            const continuation = yield this.getContinuationData();
            return new Search(this.actions, continuation, true);
        });
    }
}
exports.default = Search;
